design/external-processing-design.md

```diff
@@ -0,0 +1,648 @@
+
+- [#1015][1] Enable Lua filter
+- [#1176][2] internal/grpc: properly handle Envoy ACK/NACK protocol
+- [#2385][3] Add proxy WASM support
+- [#2475][4] Need Envoy http filter in contour
+- [#3006][5] Support for Envoy Filters (Particularly Lua)
+- [#4276][6] Exploration: WASM in Contour
+- [#5038][7] Support for External Processing Filter
+- [#5123][8] [Feature Discussion] The ways to extend/custom functionalities for Envoy: WASM vs. Lua vs. External Processing vs. GO filter.
+
+## Goals
+
+- Stick to positioning as an escape hatch.
+- Support Envoy's L7 External Processing filter for HTTP(s) Virtual Hosts.
+- Allow operators to integrate existing custom HTTP message processing service with Contour.
+- Decouple Contour from external processing, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+- Supporting set **at most one** External Processing Service for HTTP(s) endpoints.
+- Supporting set **at most one** the Global External Processing Service for Virtual Hosts.
+
+## Non Goals
+
+- Abstracting the Envoy external HTTP message processing mechanism.
+  The scope of abstracting the Envoy external processing mechanism is too large to be tractable.
+  Abstracting the protocol would also work against the goal of being able to integrate existing HTTP message processing servers.
+
+## High-Level Design
+
+Contour will add HTTP support for Envoy's External Processing.
+
+new type: `ExternalProcessor` and its friends: `ExtProc`, `ExtProcOverride`, `ProcessingMode`, `HeaderMutationRules`,`GRPCService`, `ExtProcPolicy` and more, will be defined for implement the design.
+
+In this design, the configuration is divided into three levels: `Global`, `VirtualHost`, `Route`, each level can be set up to one External processing; each level has a `disabled` option, but at different levels, it has different meanings.
+
+If the external procssing is added to the filter chain(s), it will be inserted just before the `Router` filter.
+
+### Global level
+
+At the `Global` level, there is at most one external processing configured, and if the `globalExtProc` is NOT nil, and the `processor` is set but `disabled == false`, then it will be append to the filter chain for HTTP and the default chain for HTTPS if this VirtualHost has enabled the fallback certificate; but for the normal HTTPS it varies depending on the configuration at the `VirtualHost` level(see below). If `disabled == true`, it will be ignored.
```

Comment:
why do we need `disabled == false`? cant we check the for the existence  of `globalExtProc` to make a decision whether to enable or disable `globalExtProc`?

Reply:
Just for convenience, because the extProc configuration item is a bit complicated, so that in the test or some other situations, you do not have to repeatedly add and remove the corresponding extProc, for whether to remove it, my attitude is open.